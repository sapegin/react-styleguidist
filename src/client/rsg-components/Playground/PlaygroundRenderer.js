import React from 'react';
import PropTypes from 'prop-types';
import cx from 'clsx';
import Styled from 'rsg-components/Styled';

export const styles = ({ space, color, borderRadius }) => ({
	root: {
		marginBottom: space[4],
	},
	preview: {
		padding: space[2],
		border: [[1, color.border, 'solid']],
		borderRadius,
		// the next 2 lines are required to contain floated components
		width: '100%',
		display: 'inline-block',
	},
	controls: {
		display: 'flex',
		alignItems: 'center',
		marginBottom: space[1],
	},
	toolbar: {
		marginLeft: 'auto',
	},
	tab: {}, // expose className to allow using it in 'styles' settings
	padded: {
		// add padding between each example element rendered
		'& > *': {
			isolate: false,
			marginLeft: -space[1],
			marginRight: -space[1],
			'& > *': {
				isolate: false,
				marginRight: space[1],
				marginLeft: space[1],
			},
		},
	},
});

export function PlaygroundRenderer({
	classes,
	exampleIndex,
	name,
	padded,
	preview,
	previewProps,
	tabButtons,
	tabBody,
	toolbar,
}) {
	const { className, ...props } = previewProps;
	const previewClasses = cx(classes.preview, className, { [classes.padded]: padded });
	return (
		<div className={classes.root} data-testid={`${name}-example-${exampleIndex}`}>
			<div className={previewClasses} {...props} data-preview={name} data-testid="preview-wrapper">
				{preview}
			</div>
			<div className={classes.controls}>
				<div className={classes.tabs}>{tabButtons}</div>
				<div className={classes.toolbar}>{toolbar}</div>
			</div>
			<div className={classes.tab}>{tabBody}</div>
		</div>
	);
}

PlaygroundRenderer.propTypes = {
	classes: PropTypes.object.isRequired,
	exampleIndex: PropTypes.number.isRequired,
	name: PropTypes.string.isRequired,
	padded: PropTypes.bool.isRequired,
	preview: PropTypes.node.isRequired,
	previewProps: PropTypes.object.isRequired,
	tabButtons: PropTypes.node.isRequired,
	tabBody: PropTypes.node.isRequired,
	toolbar: PropTypes.node.isRequired,
};

export default Styled(styles)(PlaygroundRenderer);
