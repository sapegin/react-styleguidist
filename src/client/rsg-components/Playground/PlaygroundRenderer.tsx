import React from 'react';
import PropTypes from 'prop-types';
import cx from 'clsx';
import Styled, { JssInjectedProps } from 'rsg-components/Styled';
import * as Rsg from '../../../typings';

export const styles = ({ space, color, borderRadius }: Rsg.Theme) => ({
	root: {
		marginBottom: space[4],
	},
	preview: {
		padding: space[2],
		border: [[1, color.border, 'solid']],
		borderRadius,
		// the next 2 lines are required to contain floated components
		width: '100%',
		display: 'inline-block',
	},
	controls: {
		display: 'flex',
		alignItems: 'center',
		marginBottom: space[1],
	},
	toolbar: {
		marginLeft: 'auto',
	},
	tab: {}, // expose className to allow using it in 'styles' settings
	padded: {
		// add padding between each example element rendered
		'& > *': {
			isolate: false,
			marginLeft: -space[1],
			marginRight: -space[1],
			'& > *': {
				isolate: false,
				marginRight: space[1],
				marginLeft: space[1],
			},
		},
	},
});

interface PlaygroundRendererProps extends JssInjectedProps {
	exampleIndex: number | string;
	componentName: string;
	padded: boolean;
	preview: React.ReactNode;
	previewClassName?: string;
	tabButtons: React.ReactNode;
	tabBody: React.ReactNode;
	toolbar: React.ReactNode;
}

export const PlaygroundRenderer: React.FunctionComponent<PlaygroundRendererProps> = ({
	classes,
	exampleIndex,
	componentName,
	padded,
	preview,
	previewClassName,
	tabButtons,
	tabBody,
	toolbar,
}) => {
	const previewClasses = cx(classes.preview, previewClassName, { [classes.padded]: padded });
	return (
		<div className={classes.root} data-testid={`${componentName}-example-${exampleIndex}`}>
			<div className={previewClasses} data-preview={componentName} data-testid="preview-wrapper">
				{preview}
			</div>
			<div className={classes.controls}>
				<div className={classes.tabs}>{tabButtons}</div>
				<div className={classes.toolbar}>{toolbar}</div>
			</div>
			<div className={classes.tab}>{tabBody}</div>
		</div>
	);
};

PlaygroundRenderer.propTypes = {
	classes: PropTypes.objectOf(PropTypes.string.isRequired).isRequired,
	exampleIndex: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
	componentName: PropTypes.string.isRequired,
	padded: PropTypes.bool.isRequired,
	preview: PropTypes.node.isRequired,
	previewClassName: PropTypes.string,
	tabButtons: PropTypes.node.isRequired,
	tabBody: PropTypes.node.isRequired,
	toolbar: PropTypes.node.isRequired,
};

export default Styled<PlaygroundRendererProps>(styles)(PlaygroundRenderer);
