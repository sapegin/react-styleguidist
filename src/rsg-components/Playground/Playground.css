.root {
	composes: border from "../../styles/colors.css";
	margin-bottom: 30px;
	border-radius: 3px;
}

.editor {
}

.preview {
	margin-bottom: 3px;
	padding: 15px;
}

.codeToggle {
	composes: font from "../../styles/common.css";
	composes: border from "../../styles/colors.css";
	composes: link from "../../styles/colors.css";
	cursor: pointer;
	display: inline-block;
	float: right;
	margin-bottom: -30px;
	margin-right: -1px;
	border-bottom-left-radius: 4px;
	border-bottom-right-radius: 4px;
	border-top: 0;
	padding: 4px 8px;
	font-size: small;
}

.showCode {
	composes: codeToggle;
	background: #fafafa;
}

.hideCode {
	composes: codeToggle;
	background: #f5f5f5;
}
