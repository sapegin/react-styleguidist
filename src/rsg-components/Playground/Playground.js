import React, { Component, PropTypes } from 'react';
import Editor from 'rsg-components/Editor';
import Preview from 'rsg-components/Preview';

import s from './Playground.css';

export default class Playground extends Component {
	static propTypes = {
		highlightTheme: PropTypes.string.isRequired,
		code: PropTypes.string.isRequired,
		evalInContext: PropTypes.func.isRequired,
		showCode: PropTypes.bool,
	};

	constructor(props) {
		super();
		this.state = {
			code: props.code,
			showCode: false,
		};
	}

	componentWillReceiveProps(nextProps) {
		let { code } = nextProps;
		if (code) {
			this.setState({
				code,
			});
		}
	}

	handleChange(newCode) {
		this.setState({
			code: newCode,
		});
	}

	handleCodeToggle() {
		this.setState({
			showCode: !this.state.showCode,
		});
	}

	render() {
		let { code } = this.state;
		let { highlightTheme } = this.props;

		return (
			<div className={s.root}>
				<div className={s.preview}>
					<Preview code={code} evalInContext={this.props.evalInContext} />
				</div>
				{this.state.showCode ? (
					<div className={s.editor}>
						<Editor code={code} highlightTheme={highlightTheme} onChange={code => this.handleChange(code)} />
						<div className={s.hideCode} onClick={() => this.handleCodeToggle()}>
							hide code
						</div>
					</div>
				) : (
					<div className={s.showCode} onClick={() => this.handleCodeToggle()}>
						show code
					</div>
				)}
			</div>
		);
	}
}
