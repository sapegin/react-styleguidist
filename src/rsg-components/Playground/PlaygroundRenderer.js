import React, { PropTypes } from 'react';
import Editor from 'rsg-components/Editor';
import Link from 'rsg-components/Link';
import Preview from 'rsg-components/Preview';
import Styled from 'rsg-components/Styled';

const styles = ({ font, link, linkHover, border, baseBackground, codeBackground }) => ({
	root: {
		position: 'relative',
		marginBottom: 30,
		border: [[1, border, 'solid']],
		borderRadius: '3px 3px 0 3px',
		'&:hover $isolatedLink': {
			isolate: false,
			opacity: 1,
		},
	},
	preview: {
		marginBottom: 3,
		padding: 15,
	},
	codeToggle: {
		position: 'absolute',
		right: -1,
		margin: 0,
		padding: [[6, 8]],
		fontFamily: font,
		fontSize: 14,
		lineHeight: 1,
		color: link,
		border: [[1, border, 'solid']],
		borderTop: 0,
		borderBottomLeftRadius: 3,
		borderBottomRightRadius: 3,
		cursor: 'pointer',
		'&:hover, &:active': {
			isolate: false,
			color: linkHover,
		},
	},
	showCode: {
		composes: '$codeToggle',
		backgroundColor: baseBackground,
	},
	hideCode: {
		composes: '$codeToggle',
		backgroundColor: codeBackground,
	},
	isolatedLink: {
		position: 'absolute',
		top: 0,
		right: 0,
		padding: [[6, 8]],
		fontFamily: font,
		fontSize: 14,
		opacity: 0,
		transition: 'opacity ease-in-out .15s .2s',
	},
});

export function PlaygroundRenderer({
	classes,
	code,
	showCode,
	name,
	index,
	isolatedExample,
	evalInContext,
	onChange,
	onCodeToggle,
}) {
	return (
		<div className={classes.root}>
			<div className={classes.preview} data-preview={name ? name : ''}>
				<div className={classes.isolatedLink}>
					{name && (
						isolatedExample ? (
							<Link href={'#!/' + name}>⇽ Exit Isolation</Link>
						) : (
							<Link href={'#!/' + name + '/' + index}>Open isolated ⇢</Link>
						)
					)}
				</div>
				<Preview code={code} evalInContext={evalInContext} />
			</div>
			{showCode ? (
				<div>
					<Editor code={code} onChange={onChange} />
					<button type="button" className={classes.hideCode} onClick={onCodeToggle}>
						Hide code
					</button>
				</div>
			) : (
				<button type="button" className={classes.showCode} onClick={onCodeToggle}>
					Show code
				</button>
			)}
		</div>
	);
}

PlaygroundRenderer.propTypes = {
	classes: PropTypes.object.isRequired,
	code: PropTypes.string.isRequired,
	showCode: PropTypes.bool.isRequired,
	index: PropTypes.number.isRequired,
	evalInContext: PropTypes.func.isRequired,
	onChange: PropTypes.func.isRequired,
	onCodeToggle: PropTypes.func.isRequired,
	name: PropTypes.string,
	isolatedExample: PropTypes.bool,
};

export default Styled(styles)(PlaygroundRenderer);
