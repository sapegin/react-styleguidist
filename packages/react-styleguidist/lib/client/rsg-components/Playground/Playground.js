import "core-js/modules/es7.symbol.async-iterator";
import "core-js/modules/es6.symbol";
import "core-js/modules/es6.function.name";

import React, { Component } from 'react';
import { polyfill } from 'react-lifecycles-compat';
import PropTypes from 'prop-types';
import debounce from 'lodash/debounce';
import Preview from 'rsg-components/Preview';
import Para from 'rsg-components/Para';
import Slot from 'rsg-components/Slot';
import PlaygroundRenderer from 'rsg-components/Playground/PlaygroundRenderer';
import { EXAMPLE_TAB_CODE_EDITOR } from '../slots';
import { DisplayModes, ExampleModes } from '../../consts';

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (let i = 0; i < props.length; i++) { const descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) {descriptor.writable = true;} Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) {_defineProperties(Constructor.prototype, protoProps);} if (staticProps) {_defineProperties(Constructor, staticProps);} return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) {_setPrototypeOf(subClass, superClass);} }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const Playground =
/*#__PURE__*/
function (_Component) {
  _inherits(Playground, _Component);

  function Playground(props, context) {
    let _this;

    _classCallCheck(this, Playground);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Playground).call(this, props, context));
    const code = props.code;

        
const settings = props.settings;

        
const exampleMode = props.exampleMode;
    const config = context.config;
    const expandCode = exampleMode === ExampleModes.expand;
    const activeTab = settings.showcode !== undefined ? settings.showcode : expandCode;
    _this.state = {
      code,
      prevCode: code,
      activeTab: activeTab ? EXAMPLE_TAB_CODE_EDITOR : undefined
    };
    _this.handleTabChange = _this.handleTabChange.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    _this.handleChange = debounce(_this.handleChange.bind(_assertThisInitialized(_assertThisInitialized(_this))), config.previewDelay);
    return _this;
  }

  _createClass(Playground, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      // Clear pending changes
      this.handleChange.cancel();
    }
  }, {
    key: "handleChange",
    value: function handleChange(code) {
      this.setState({
        code
      });
    }
  }, {
    key: "handleTabChange",
    value: function handleTabChange(name) {
      this.setState(function (state) {
        return {
          activeTab: state.activeTab !== name ? name : undefined
        };
      });
    }
  }, {
    key: "render",
    value: function render() {
      const _this$state = this.state;

          
const code = _this$state.code;

          
const activeTab = _this$state.activeTab;
      const _this$props = this.props;

          
const evalInContext = _this$props.evalInContext;

          
const index = _this$props.index;

          
const name = _this$props.name;

          
const settings = _this$props.settings;

          
const exampleMode = _this$props.exampleMode;
      const displayMode = this.context.displayMode;
      const isExampleHidden = exampleMode === ExampleModes.hide;
      const isEditorHidden = settings.noeditor || isExampleHidden;
      const preview = React.createElement(Preview, {
        code,
        evalInContext
      });
      return isEditorHidden ? React.createElement(Para, null, preview) : React.createElement(PlaygroundRenderer, {
        name,
        preview,
        previewProps: settings.props || {},
        tabButtons: React.createElement(Slot, {
          name: "exampleTabButtons",
          active: activeTab,
          props: {
            onClick: this.handleTabChange
          }
        }),
        tabBody: React.createElement(Slot, {
          name: "exampleTabs",
          active: activeTab,
          onlyActive: true // evalInContext passed through to support custom slots that eval code
          ,
          props: {
            code,
            onChange: this.handleChange,
            evalInContext
          }
        }),
        toolbar: React.createElement(Slot, {
          name: "exampleToolbar",
          props: {
            name,
            isolated: displayMode === DisplayModes.example,
            example: index
          }
        })
      });
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      const code = nextProps.code;

      if (prevState.prevCode !== code) {
        return {
          prevCode: code,
          code
        };
      }

      return null;
    }
  }]);

  return Playground;
}(Component);

_defineProperty(Playground, "propTypes", {
  code: PropTypes.string.isRequired,
  evalInContext: PropTypes.func.isRequired,
  index: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  exampleMode: PropTypes.string.isRequired,
  settings: PropTypes.object
});

_defineProperty(Playground, "contextTypes", {
  config: PropTypes.object.isRequired,
  displayMode: PropTypes.string
});

export default polyfill(Playground);