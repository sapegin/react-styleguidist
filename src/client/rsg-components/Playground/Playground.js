import React, { Component } from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash/debounce';
import Preview from 'rsg-components/Preview';
import Para from 'rsg-components/Para';
import Slot from 'rsg-components/Slot';
import PlaygroundRenderer from 'rsg-components/Playground/PlaygroundRenderer';
import Context from 'rsg-components/Context';
import { EXAMPLE_TAB_CODE_EDITOR } from '../slots';
import { DisplayModes, ExampleModes } from '../../consts';

class Playground extends Component {
	static propTypes = {
		code: PropTypes.string.isRequired,
		evalInContext: PropTypes.func.isRequired,
		index: PropTypes.number.isRequired,
		name: PropTypes.string.isRequired,
		exampleMode: PropTypes.string.isRequired,
		settings: PropTypes.object,
	};
	static defaultProps = {
		settings: {},
	};
	static contextType = Context;

	handleChange = debounce(code => {
		this.setState({
			code,
		});
	}, this.context.config.previewDelay);

	state = {
		code: this.props.code,
		prevCode: this.props.code,
		activeTab: this.getInitialActiveTab() ? EXAMPLE_TAB_CODE_EDITOR : undefined,
	};

	static getDerivedStateFromProps(nextProps, prevState) {
		const { code } = nextProps;
		if (prevState.prevCode !== code) {
			return {
				prevCode: code,
				code,
			};
		}
		return null;
	}

	componentWillUnmount() {
		// Clear pending changes
		this.handleChange.cancel();
	}

	getInitialActiveTab() {
		const expandCode = this.props.exampleMode === ExampleModes.expand;
		return this.props.settings.showcode !== undefined ? this.props.settings.showcode : expandCode;
	}

	handleTabChange = name => {
		this.setState(state => ({
			activeTab: state.activeTab !== name ? name : undefined,
		}));
	};

	render() {
		const { code, activeTab } = this.state;
		const { evalInContext, index, name, settings, exampleMode } = this.props;
		const { displayMode } = this.context;
		const isExampleHidden = exampleMode === ExampleModes.hide;
		const isEditorHidden = settings.noeditor || isExampleHidden;
		const preview = <Preview code={code} evalInContext={evalInContext} />;

		return isEditorHidden ? (
			<Para>{preview}</Para>
		) : (
			<PlaygroundRenderer
				name={name}
				exampleIndex={index}
				padded={!!settings.padded}
				preview={preview}
				previewProps={settings.props || {}}
				tabButtons={
					<Slot
						name="exampleTabButtons"
						active={activeTab}
						props={{ onClick: this.handleTabChange }}
					/>
				}
				tabBody={
					<Slot
						name="exampleTabs"
						active={activeTab}
						onlyActive
						// evalInContext passed through to support custom slots that eval code
						props={{ code, onChange: this.handleChange, evalInContext }}
					/>
				}
				toolbar={
					<Slot
						name="exampleToolbar"
						props={{ name, isolated: displayMode === DisplayModes.example, example: index }}
					/>
				}
			/>
		);
	}
}

export default Playground;
