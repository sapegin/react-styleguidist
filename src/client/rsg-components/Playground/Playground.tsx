import React, { Component } from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash/debounce';
import Preview from 'rsg-components/Preview';
import Para from 'rsg-components/Para';
import Slot from 'rsg-components/Slot';
import PlaygroundRenderer from 'rsg-components/Playground/PlaygroundRenderer';
import Context from 'rsg-components/Context';
import { EXAMPLE_TAB_CODE_EDITOR } from '../slots';
import { getIsolatedUrl } from '../../utils/getUrl';
import { ExampleModes } from '../../consts';
import * as Rsg from '../../../typings';

const getIsolatedButtonUrl = ({
	isolated,
	hashPath,
	exampleIndex,
}: {
	isolated: boolean;
	hashPath: string[];
	exampleIndex?: number | string;
}) => {
	if (isolated) {
		return getIsolatedUrl(hashPath);
	} else {
		return getIsolatedUrl(hashPath, exampleIndex);
	}
};

interface PlaygroundProps {
	componentName: string;
	componentHashPath: string[];
	exampleMode: string;
	code: string;
	documentScope: Record<string, unknown>;
	exampleScope: Record<string, unknown>;
	modifiers: Rsg.Modifiers;
}

interface PlaygroundState {
	code: string;
	prevCode: string;
	activeTab?: string;
}

class Playground extends Component<PlaygroundProps, PlaygroundState> {
	public static propTypes = {
		componentName: PropTypes.string.isRequired,
		componentHashPath: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
		exampleMode: PropTypes.string.isRequired,
		code: PropTypes.string.isRequired,
		documentScope: PropTypes.object.isRequired,
		exampleScope: PropTypes.object.isRequired,
		modifiers: PropTypes.object.isRequired,
	};

	public static contextType = Context;

	private handleChange = debounce((code) => {
		this.setState({
			code,
		});
	}, this.context.config.previewDelay);

	public state: PlaygroundState = {
		code: this.props.code,
		prevCode: this.props.code,
		activeTab: this.getInitialActiveTab() ? EXAMPLE_TAB_CODE_EDITOR : undefined,
	};

	public static getDerivedStateFromProps(nextProps: PlaygroundProps, prevState: PlaygroundState) {
		const { code } = nextProps;
		if (prevState.prevCode !== code) {
			return {
				prevCode: code,
				code,
			};
		}
		return null;
	}

	public componentWillUnmount() {
		// Clear pending changes
		this.handleChange.cancel();
	}

	private getInitialActiveTab(): boolean {
		const { showcode } = this.props.modifiers;
		const expandCode = this.props.exampleMode === ExampleModes.expand;
		return showcode !== undefined ? showcode : expandCode;
	}

	private handleTabChange = (name: string) => {
		this.setState((state) => ({
			activeTab: state.activeTab !== name ? name : undefined,
		}));
	};

	public render() {
		const { code, activeTab } = this.state;
		const {
			documentScope,
			exampleScope,
			componentName,
			componentHashPath,
			modifiers,
			exampleMode,
		} = this.props;
		const { isolated, exampleIndex } = this.context;
		const isExampleHidden = exampleMode === ExampleModes.hide;
		const isEditorHidden = modifiers.noeditor || isExampleHidden;
		const isolatedButton = isolated && exampleIndex !== undefined;

		const preview = (
			<Preview code={code} documentScope={documentScope} exampleScope={exampleScope} />
		);

		return isEditorHidden ? (
			<Para>{preview}</Para>
		) : (
			<PlaygroundRenderer
				componentName={componentName}
				exampleIndex={modifiers.index}
				padded={!!modifiers.padded}
				preview={preview}
				previewClassName={modifiers['preview-class']}
				tabButtons={
					<Slot
						name="exampleTabButtons"
						active={activeTab}
						props={{ onClick: this.handleTabChange }}
					/>
				}
				tabBody={
					<Slot
						name="exampleTabs"
						active={activeTab}
						onlyActive
						props={{ code, onChange: this.handleChange }}
					/>
				}
				toolbar={
					<Slot
						name="exampleToolbar"
						props={{
							name: componentName,
							href: getIsolatedButtonUrl({
								isolated: isolatedButton,
								hashPath: componentHashPath,
								exampleIndex: modifiers.index,
							}),
							exampleIndex: modifiers.index,
							isolated: isolatedButton,
						}}
					/>
				}
			/>
		);
	}
}

export default Playground;
