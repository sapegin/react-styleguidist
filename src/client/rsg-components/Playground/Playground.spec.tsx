import React from 'react';
import { render, fireEvent } from '../../test';
import Playground from './Playground';

const code = 'import React from "react"; <button>Code: OK</button>';
const newCode = 'import React from "react"; <button>Code: Not OK</button>';
const defaultProps = {
	index: 0,
	componentName: 'name',
	componentHashPath: ['Name'],
	modifiers: {
		index: 1,
	},
	exampleMode: 'collapse',
	documentScope: {},
	exampleScope: { react: React },
	code,
};

it('should update code via props', () => {
	const { rerender, getByText } = render(<Playground {...defaultProps} />);

	expect(getByText('Code: OK')).toBeInTheDocument();

	rerender(<Playground {...defaultProps} code={newCode} />);

	expect(getByText('Code: Not OK')).toBeInTheDocument();
});

it('should open a code editor', () => {
	const { queryByRole, getByText } = render(<Playground {...defaultProps} />);

	expect(queryByRole('textbox')).not.toBeInTheDocument();

	fireEvent.click(getByText(/view code/i));

	expect(queryByRole('textbox')).toBeInTheDocument();
});

it('should not render a code editor if noeditor option passed in example settings', () => {
	const { queryByText } = render(
		<Playground {...defaultProps} modifiers={{ ...defaultProps.modifiers, noeditor: true }} />
	);

	expect(queryByText(/view code/i)).not.toBeInTheDocument();
});

it('should open a code editor by default if showcode=true option passed in example settings', () => {
	const { queryByRole } = render(
		<Playground {...defaultProps} modifiers={{ ...defaultProps.modifiers, showcode: true }} />
	);

	expect(queryByRole('textbox')).toBeInTheDocument();
});

it('should open a code editor by default if exampleMode="expand" option specified in style guide config', () => {
	const { queryByRole } = render(<Playground {...defaultProps} exampleMode="expand" />);

	expect(queryByRole('textbox')).toBeInTheDocument();
});

it('showcode option in example settings should overwrite style guide config option', () => {
	const { queryByRole } = render(
		<Playground
			{...defaultProps}
			exampleMode="expand"
			modifiers={{ ...defaultProps.modifiers, showcode: false }}
		/>
	);

	expect(queryByRole('textbox')).not.toBeInTheDocument();
});

it('should not include padded class if padded option is not passed in example settings', () => {
	const { getByTestId } = render(
		<Playground {...defaultProps} modifiers={{ ...defaultProps.modifiers, padded: false }} />
	);

	expect(getByTestId('preview-wrapper')).not.toHaveAttribute(
		'class',
		expect.stringContaining('rsg--padded-')
	);
});

it('should include padded class if padded option is passed in example settings', () => {
	const { getByTestId } = render(
		<Playground {...defaultProps} modifiers={{ ...defaultProps.modifiers, padded: true }} />
	);

	expect(getByTestId('preview-wrapper')).toHaveAttribute(
		'class',
		expect.stringContaining('rsg--padded-')
	);
});
