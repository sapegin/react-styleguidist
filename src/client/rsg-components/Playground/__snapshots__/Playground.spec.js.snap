// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer should render preview 1`] = `
<div
  className="root"
  data-testid="name-example-0"
>
  <div
    className="preview pizza"
    data-preview="name"
    title="salami"
  >
    <div>
      preview
    </div>
  </div>
  <div
    className="controls"
  >
    <div>
      <div>
        tab buttons
      </div>
    </div>
    <div
      className="toolbar"
    >
      <div>
        toolbar
      </div>
    </div>
  </div>
  <div
    className="tab"
  >
    <div>
      tab body
    </div>
  </div>
</div>
`;

exports[`should render component renderer 1`] = `
<Styled(Playground)
  exampleIndex={0}
  name="name"
  preview={
    <Preview
      code=
<button>
  OK
</button>

      evalInContext={[Function]}
    />
  }
  previewProps={Object {}}
  tabBody={
    <Slot
      name="exampleTabs"
      onlyActive={true}
      props={
        Object {
          "code": 
<button>
  OK
</button>
,
          "evalInContext": [Function],
          "onChange": [Function],
        }
      }
    />
  }
  tabButtons={
    <Slot
      name="exampleTabButtons"
      props={
        Object {
          "onClick": [Function],
        }
      }
    />
  }
  toolbar={
    <Slot
      name="exampleToolbar"
      props={
        Object {
          "example": 0,
          "isolated": false,
          "name": "name",
        }
      }
    />
  }
/>
`;
